### ISSUE TITLE
Build logic should be able to mutate declared dependencies

### ISSUE THREAD

**Issue by bigdaz** [gradle/gradle#3245]
## Overview
Provide a mechanism for a plugin to provide logic that adds/removes/modifies the set of dependencies for a project configuration. This logic should be evaluated before the dependency set is used: for resolution, publishing or other purpose. The logic should evaluated only once per build execution, and should be deferred until the dependency set is actually used.

## Use cases
1. The nebula.dependency-recommender and io.spring.dependency-management plugins use the beforeResolve hook to provide versions for declared dependencies of a project. These versions should automatically be included wherever that project is consumed, whether by project dependency, composite build or published metadata.
2. Dependencies added via Configuration.defaultDependencies() are not reflected in metadata consumed for a project.
3. Build logic provides dependency information for a project that is expensive to calculate. This calculation should not be performed until dependencies are actually required for resolution or publishing.

## Solution
Provide a convenient mechanism to modify dependency declarations before they are consumed or used.  Modifications would apply for both resolved and published dependency representations, but would not affect transitive dependencies.

- Allow version to be modified on `ExternalModuleDependency`. Currently the only way to update the version is to make a deep copy and replace the entire dependency.
- Build logic has a hook where all direct dependencies can be modified before resolution or publishing
    - Transitive dependencies would not be modified by this hook
- Updated dependencies are everywhere that the first-level dependency declaration can be consumed:
    - Metadata files published with ivy-publish and maven-publish
    - Component metadata resolved for project dependencies
    - Component metadata resolved for projects substituted in a composite build
- Multiple rules will be executed in the order they are added
- Deprecate the ability to mutate a DependencySet (or any included Dependency) within a beforeResolve rule
- Configuration.defaultDependencies will not yet be deprecated. These rules run before any others, and only if the DependencySet is empty.

## DSL
```
configurations.all {
  withDependencies { DependencySet deps ->
    deps.withType(ExternalModuleDependency).each { d ->
      if (d.version == null) {
        // Plugin looks up 'recommended' version for this build
        String recommendedVersion = 
                  lookupVersion(d.requested.group, d.requested.name)
        if (recommendedVersion != null) {
          d.version = recommendedVersion
        }
      }
    }
    if (deps.empty) {
       deps.add project.dependencies.create("com.google.guava:guava:17.0")
    }
  }
}
```

## Publishing
All rules will be evaluated prior to publishing, so that the mutated dependency information appears in the published metadata.

## Story breakdown
- Allow an action to mutate declared dependencies for a configuration
- Allow build logic to update the version of a declared module dependency
- Mutated dependencies are available to downstream consumers
- Un-incubate new API and deprecate the mutation of DependencySet from within Configuration.beforeResolve.

**Comment by adammurdoch** [comment]
It would be good to have some detail about what happens to the dependencies associated with detached and copied configurations when they are resolved, and how the action works with inherited dependencies. Otherwise the plan looks great.

**Comment by adammurdoch** [comment]
Also how we might apply this logic to plugins and their dependencies.

**Comment by bigdaz** [comment]
This has been done enough for now. I'm not sure that we want to retain this mechanism long term, or instead provide a way for a component-metadata rule to apply to the local (resolving) component.
